packages/sdk/ts/user/user.ts

```diff
@@ -124,3 +141,159 @@ export const isJoinedUser = async ({
+export const joinPoll = async ({
+  maciAddress,
+  privateKey,
+  stateFile,
+  pollId,
+  signer,
+  startBlock,
+  endBlock,
+  blocksPerBatch,
+  pollJoiningZkey,
+  useWasm,
+  rapidsnark,
+  pollWitgen,
+  pollWasm,
+  sgDataArg,
+  ivcpDataArg,
+}: IJoinPollArgs): Promise<IJoinPollData> => {
+  const validContract = await contractExists(signer.provider!, maciAddress);
+
+  if (!validContract) {
+    throw new Error("MACI contract does not exist");
+  }
+
+  if (!PrivKey.isValidSerializedPrivKey(privateKey)) {
+    throw new Error("Invalid MACI private key");
+  }
+
+  if (pollId < 0) {
+    throw new Error("Invalid poll id");
+  }
+
+  const userMaciPrivKey = PrivKey.deserialize(privateKey);
+  const userMaciPubKey = new Keypair(userMaciPrivKey).pubKey;
+  const nullifier = poseidon([BigInt(userMaciPrivKey.asCircuitInputs()), pollId]);
+
+  // check if the user has already joined the poll based on the nullifier
+  const hasUserJoinedAlready = await hasUserJoinedPoll({
+    maciAddress,
+    pollId,
+    nullifier,
+    signer,
+  });
+
+  if (hasUserJoinedAlready) {
+    throw new Error("User has already joined");
+  }
+
+  const maciContract = MACIFactory.connect(maciAddress, signer);
+  const pollContracts = await maciContract.getPoll(pollId);
+  const pollContract = PollFactory.connect(pollContracts.poll, signer);
+
+  // get the state index from the MACI contract
+  const stateIndex = await maciContract.getStateIndex(userMaciPubKey.hash());
+
+  let circuitInputs: CircuitInputs;
+  if (stateFile) {
+    circuitInputs = await getPollJoiningCircuitInputsFromStateFile({
+      stateFile,
+      pollId,
+      stateIndex,
+      userMaciPrivKey,
+    });
+  } else {
+    circuitInputs = await getPollJoiningCircuitEvents({
+      maciContract,
+      stateIndex,
+      pollId,
+      userMaciPrivKey,
+      signer,
+      startBlock,
+      endBlock,
+      blocksPerBatch,
+    });
+  }
+
+  const currentStateRootIndex = Number.parseInt((await maciContract.numSignUps()).toString(), 10) - 1;
+
+  try {
+    // generate the proof for this batch
+    const proof = await generateAndVerifyProof(
+      circuitInputs,
+      pollJoiningZkey,
+      useWasm,
+      rapidsnark,
+      pollWitgen,
+      pollWasm,
+    );
+
+    // submit the message onchain as well as the encryption public key
+    const tx = await pollContract.joinPoll(
+      nullifier,
+      userMaciPubKey.asContractParam(),
+      currentStateRootIndex,
+      proof,
+      sgDataArg,
+      ivcpDataArg,
+    );
+    const receipt = await tx.wait();
+
+    if (receipt?.status !== 1) {
+      throw new Error("Transaction failed");
+    }
+
+    const iface = pollContract.interface;
+
+    // get user data from the event
+    const [log] = receipt.logs;
+    /**
+     *      const event = iface.parseLog(logs as unknown as { topics: string[]; data: string }) as unknown as {
+          args: { _pollStateIndex: bigint; _voiceCreditBalance: bigint };
+        };
+     * 
+     */
```

Comment:
Let's format it

Reply:
oh actually I need to remove this it's not needed